### PR TITLE
[New] `sort-prop-types`: give errors on TS types

### PR DESCRIPTION
**Pull request by akulsr0**
This PR is an attempt to solve #3578

### CODE REVIEW

**Comment by ljharb** [comment]
Seems like there's some uncovered added lines; can you add tests that hit them?

**Comment by ljharb** [comment]
Looks like there's still a few failing tests.

**Comment by akulsr0** [comment]
Yep, tried to debug this. It seems for `node 8/9` and `eslint 6`, the `variableUtil.findVariableByName` util is unable to grab the type node. Not sure why this is happening

**Comment by akulsr0** [comment]
@ljharb I have fixed the previous failing tests.

File: lib/rules/sort-prop-types.js [hunk 1]
```diff
@@ -215,6 +217,22 @@ module.exports = {
       }
     }
 
+    function handleFunctionComponent(node) {
+      const firstArg = node.params[0].typeAnnotation && node.params[0].typeAnnotation.typeAnnotation;
+      if (firstArg && firstArg.type === 'TSTypeReference') {
+        const propType = variableUtil.findVariableByName(context, firstArg.typeName.name)
+        || (typeAnnotations.get(firstArg.typeName.name)
+          && typeAnnotations.get(firstArg.typeName.name)[0]);
```
Comment (ljharb) [inline lib/rules/sort-prop-types.js:225]:
```suggestion
          || (
            typeAnnotations.get(firstArg.typeName.name)
            && typeAnnotations.get(firstArg.typeName.name)[0]
          );
```
looks like || fallback still isn't covered by a test case?

Comment (akulsr0) [reply lib/rules/sort-prop-types.js:225]: I am not sure why it comes as uncovered, this fallback is essentially the fix for previous test case failure which was happening for `node 8/9` and `eslint 6`

Comment (ljharb) [reply lib/rules/sort-prop-types.js:225]: ah, perhaps it's codecov maxing out on what it can merge 🤔

Comment (akulsr0) [reply lib/rules/sort-prop-types.js]: Fixed it, all checks are passing now.

**Comment by ljharb** [comment]
ok, I rebased this, but I think this needs to go behind an option, otherwise it would be a breaking change.

File: lib/rules/sort-prop-types.js [hunk 2]
```diff
@@ -52,6 +52,9 @@ module.exports = {
         sortShapeProp: {
           type: 'boolean',
         },
+        allowTypescript: {
```
Comment (ljharb) [inline lib/rules/sort-prop-types.js:55]: let's call it `checkTypes`?

File: tests/lib/rules/sort-prop-types.js [hunk 3]
```diff
@@ -480,6 +480,19 @@ ruleTester.run('sort-prop-types', rule, {
...
+      code: `
+        type Props = {
+          zzz: string;
+          aaa: string;
+        }
+        function Foo(props: Props) {
+          return null;
+        }
+      `,
+      features: ['ts', 'no-babel'],
```
Comment (ljharb) [inline tests/lib/rules/sort-prop-types.js:494]:
```suggestion
      features: ['types'],
```
this should work for flow too, no?

File: lib/rules/sort-prop-types.js (hunk 2 above)
Comment (akulsr0) [reply lib/rules/sort-prop-types.js:55]: ok sure

File: tests/lib/rules/sort-prop-types.js (hunk 3 above)
Comment (akulsr0) [reply tests/lib/rules/sort-prop-types.js:494]: `'no-babel'` would be required.

Comment (ljharb) [reply tests/lib/rules/sort-prop-types.js:494]: why? babel is what parses flow.

File: lib/rules/sort-prop-types.js [hunk 4]
```diff
@@ -261,6 +285,29 @@ module.exports = {
...
+          const currentNode = [].concat(
+            typeAnnotations.get(node.id.name) || [],
+            node.typeAnnotation
+          );
+          typeAnnotations.set(node.id.name, currentNode);
+        }
+      },
+
+      FunctionDeclaration: handleFunctionComponent,
+      ArrowFunctionExpression: handleFunctionComponent,
```
Comment (ljharb) [inline lib/rules/sort-prop-types.js:310]: maybe:
```suggestion
      FunctionDeclaration: checkTypes && handleFunctionComponent,
      ArrowFunctionExpression: checkTypes && handleFunctionComponent,
```

File: lib/rules/sort-prop-types.js [hunk 5]
```diff
@@ -215,6 +221,24 @@ module.exports = {
       }
     }
 
+    function handleFunctionComponent(node) {
+      if (!checkTypes) {
+        return;
+      }
```
Comment (ljharb) [inline lib/rules/sort-prop-types.js:227]: instead of:
```suggestion
```

File: tests/lib/rules/sort-prop-types.js (hunk 3 above)
Comment (akulsr0) [reply tests/lib/rules/sort-prop-types.js:494]: The error type for both case is different (`TSPropertySignature` and `ObjectTypeProperty`), which I have to mention in test case.

Comment (akulsr0) [reply tests/lib/rules/sort-prop-types.js:494]: Oops, my bad mentioning the error type isn't mandatory. Fixed it.

File: lib/rules/sort-prop-types.js (hunk 4 above)
Comment (akulsr0) [reply lib/rules/sort-prop-types.js:333]: Won't this give an error if `checkTypes` is falsy?

Comment (akulsr0) [reply lib/rules/sort-prop-types.js:333]: Both (`checkTypes && handleFunctionComponent` and `checkTypes ? handleFunctionComponent : undefined`) of the above gives error, is following fine?
```js
FunctionDeclaration: checkTypes ? handleFunctionComponent : () => {},
ArrowFunctionExpression: checkTypes ? handleFunctionComponent : () => {},
```

Comment (akulsr0) [reply lib/rules/sort-prop-types.js:330]: Doing the `Object.assign` way

Comment (akulsr0) [reply lib/rules/sort-prop-types.js:330]: @ljharb this was done!

File: lib/rules/sort-prop-types.js [hunk 6]
```diff
@@ -261,6 +295,39 @@ module.exports = {
...
+        if (node.right.type === 'ObjectTypeAnnotation') {
+          const currentNode = [].concat(
+            typeAnnotations.get(node.id.name) || [],
+            node.right
+          );
+          typeAnnotations.set(node.id.name, currentNode);
+        }
+      },
+
+      TSTypeAliasDeclaration(node) {
```
Comment (ljharb) [inline lib/rules/sort-prop-types.js:319]: i think all of these can also be moved to the "checkTypes" branch below?

File: tests/lib/rules/sort-prop-types.js (hunk 3 above)
Comment (ljharb) [inline tests/lib/rules/sort-prop-types.js:494]:
```suggestion
      features: ['types'],
```

Comment (akulsr0) [reply tests/lib/rules/sort-prop-types.js]: this was done

File: lib/rules/sort-prop-types.js (hunk 2 above)
Comment (akulsr0) [reply lib/rules/sort-prop-types.js]: this was done

**Review by ljharb — APPROVED** [review]

**Pull request merged**